okio/src/commonMain/kotlin/okio/FileHandle.kt

```diff
@@ -0,0 +1,219 @@
+    }
+    return FileHandleSource(this, offset)
+  }
+
+  /**
+   * Returns the position of [source] in the file. The argument [source] must be either a source
+   * produced by this file handle, or a [BufferedSource] that directly wraps such a source. If the
+   * parameter is a [BufferedSource], it adjusts for buffered bytes.
+   */
+  @Throws(IOException::class)
+  fun position(source: Source): Long {
+    var source = source
+    var bufferSize = 0L
+
+    if (source is RealBufferedSource) {
+      bufferSize = source.buffer.size
+      source = source.source
+    }
+
+    require(source is FileHandleSource && source.fileHandle === this) {
+      "source was not created by this FileHandle"
+    }
+
+    return source.position - bufferSize
+  }
+
+  /**
+   * Returns a sink that writes to this starting at [offset]. The returned sink must be closed when
+   * it is no longer needed.
+   */
+  @Throws(IOException::class)
+  fun sink(offset: Long = 0L): Sink {
+    synchronized(this) {
+      check(!closed) { "closed" }
+      openStreamCount++
+    }
+    return FileHandleSink(this, offset)
+  }
+
+  /**
+   * Returns a sink that writes to this starting at the end. The returned sink must be closed when
+   * it is no longer needed.
+   */
+  @Throws(IOException::class)
+  fun appendingSink(): Sink {
+    return sink(size())
+  }
+
+  /**
+   * Returns the position of [sink] in the file. The argument [sink] must be either a sink produced
+   * by this file handle, or a [BufferedSink] that directly wraps such a sink. If the parameter is a
+   * [BufferedSink], it adjusts for buffered bytes.
+   */
+  @Throws(IOException::class)
+  fun position(sink: Sink): Long {
+    var sink = sink
+    var bufferSize = 0L
+
+    if (sink is RealBufferedSink) {
+      bufferSize = sink.buffer.size
+      sink = sink.sink
+    }
+
+    require(sink is FileHandleSink && sink.fileHandle === this) {
+      "sink was not created by this FileHandle"
+    }
+
+    return sink.position + bufferSize
+  }
+
+  @Throws(IOException::class)
+  override fun close() {
+    synchronized(this) {
+      if (closed) return@close
+      closed = true
+      if (openStreamCount != 0) return@close
+    }
+    closeInternal()
+  }
+
+  /**
+   * Subclasses should implement this to release resources held by this file handle. It is invoked
+   * once both the file handle is closed, and also all sources and sinks produced by it are also
+   * closed.
+   */
+  @Throws(IOException::class)
+  protected abstract fun closeInternal()
+
+  private class FileHandleSink(
+    val fileHandle: FileHandle,
+    var position: Long
+  ) : Sink {
+    var closed = false
+
+    override fun write(source: Buffer, byteCount: Long) {
+      fileHandle.write(position, source, byteCount)
+      position += byteCount
+    }
+
+    override fun flush() {
+      fileHandle.flush()
+    }
+
+    override fun timeout() = Timeout.NONE
+
+    override fun close() {
+      synchronized(this) {
```

Comment:
Is synchronizing on the sink here correct? Couldn't that race if two sinks or sources are closed at same time?

Reply:
Totally incorrect. Great catch.

Reply:
Fixed to synchronize on `fileHandle`.

Great catch. Hooray for your careful code review.